### PR TITLE
When exporting fails, check if input hash is good

### PR DESCRIPTION
**Pull request by henriquegemignani**
image

### CODE REVIEW

**Comment by Miepee** [comment]
Not sure if its the intended state, but tests are failing.
(Will review a bit later)

File: randovania/cli/commands/hash_input.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+from typing import TYPE_CHECKING
+
+from randovania.cli.cli_lib import add_debug_argument
+from randovania.exporter import game_exporter
+
+if TYPE_CHECKING:
+    from argparse import ArgumentParser, Namespace, _SubParsersAction
+
+
+def hash_input_command_logic(args: Namespace) -> int:
```
[the remark this answers is not in the document]
Comment (duncathan) [reply randovania/cli/commands/hash_input.py:13]: isn't this an exit code?

[the remark this answers is not in the document]
Comment (Miepee) [reply randovania/cli/commands/hash_input.py:13]: I thought exit codes return 0 by default if not specified?

[the remark this answers is not in the document]
Comment (henriquegemignani) [reply randovania/cli/commands/hash_input.py:13]: It's our api for these logic functions, they return a status code. Not sure it's ever not 0, but whatever

File: randovania/games/common/prime_family/exporter/good_hashes.py [hunk 2]
```diff
@@ -0,0 +1,12 @@
+PRIME1_GC_ISOS = (
+    "sha1:ac20c744db18fdf0339f37945e880708fd317231",  # NTSC 0-00
+    "sha1:4ba8933499e0b74b2f6006d622e4fbc7593ab3c7",  # NTSC 0-01
+    "sha1:1a737910b55b59c6ad91be9e3e3c43517fd52efb",  # NTSC 0-02
+    "sha1:34ac8a764a3c1db3326c39071cee2fc49e730aca",  # PAL
+    "sha1:ee6c58b46012ebffb615506ec43e02ba71905662",  # Japan
```
[the remark this answers is not in the document]
Comment (henriquegemignani) [reply randovania/games/common/prime_family/exporter/good_hashes.py:6]: I dunno, I took the most common ones.

File: randovania/exporter/game_exporter.py [hunk 3]
```diff
@@ -100,19 +100,45 @@ def export_game(
...
-                            "input_hash": input_hash,
-                        },
-                    )
-                    # TODO: compare `input_hash` with known good values and if different,
-                    # wrap into a "User Error" exception
-                    raise
+                    incorrect_hashes = self.get_unknown_input_hashes(input_hash)
+                    if incorrect_hashes:
+                        msg = "An error occurred while exporting.\nThe following files have invalid hashes:\n"
+                        msg += "\n".join(f"{key} => {value}" for key, value in incorrect_hashes.items())
```
[the remark this answers is not in the document]
Comment (henriquegemignani) [reply randovania/exporter/game_exporter.py:106]: Your error message is definitely friendlier, but both also do end up in the same situation imo: user will need to ask for help.

[the remark this answers is not in the document]
Comment (Miepee) [reply randovania/exporter/game_exporter.py:106]: for most people yeah. Some probably get the hint and will redump but I don't think its gonna be anywhere in the majority.

[the remark this answers is not in the document]
Comment (henriquegemignani) [reply randovania/exporter/game_exporter.py:106]: Probably the best message is just linking to a guide in randovania.org, and mention the key so they'll be able to find the proper instructions.

[the remark this answers is not in the document]
Comment (Miepee) [reply randovania/exporter/game_exporter.py:106]: Considering the amount of recent "why is randovania crashing, my nkit totally is a good dump" messages we've gotten, we may just wanna slap in a TODO for now to get it merged.

File: randovania/exporter/game_exporter.py [hunk 4]
```diff
@@ -100,19 +100,45 @@ def export_game(
...
+                            },
+                        )
+                        raise
 
                 finally:
                     self._after_export()
 
+    def known_good_hashes(self) -> dict[str, tuple[str, ...]]:
+        """
+        :return: A dict mapping keys returned by GameExportParams.calculate_input_hash to list of known good hashes.
```
[the remark this answers is not in the document]
Comment (Miepee) [reply randovania/exporter/game_exporter.py:122]: the "to" still sounds weird to me there. 
Feel like "to a" sounds nicer.
But its probably also insignificant

**Comment by henriquegemignani** [comment]
Updated error message:
image

File: randovania/exporter/game_exporter.py [hunk 5]
```diff
@@ -102,8 +102,12 @@ def export_game(
                     input_hash = export_params.calculate_input_hash()
                     incorrect_hashes = self.get_unknown_input_hashes(input_hash)
                     if incorrect_hashes:
-                        msg = "An error occurred while exporting.\nThe following files have invalid hashes:\n"
-                        msg += "\n".join(f"{key} => {value}" for key, value in incorrect_hashes.items())
+                        msg = "An error occurred while exporting.\n\nYour following input is invalid:\n"
```
Comment (Miepee) [inline randovania/exporter/game_exporter.py:105]: another slight nit: the title says "error during exporting" while this says "error [occured] while exporting"
would be nice to have them consistent.

**Review by Miepee — COMMENTED** [review]
Your tests are failing (working :D)
> FAILED test/games/prime1/exporter/test_prime_game_exporter.py::test_room_rando_map_maker - randovania.patching.patchers.exceptions.UnableToExportError: An error occurred while exporting.
Your following input is invalid:
> - prime1_iso
> - prime2_iso
> 
> Please check the [Input Validation](https://randovania.org/guides/input-validation) guide for more details.

File: randovania/games/common/prime_family/exporter/good_hashes.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Miepee) [reply randovania/games/common/prime_family/exporter/good_hashes.py:6]: we do support korean, if you can add a "TODO: add korean" to here, that would be nice.

**Review by Miepee — APPROVED** [review]

**Pull request merged**